### PR TITLE
Feature/batch processing

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

Copied batch processing functions from dp-frontend-filter-datase-controller, we need at least a function to get items in batches, and a function to process each batch. This has been generalised into `ProcessInConcurrentBatches`

- Implemented generic concurrent batch processing in common.go:
  - executes up to `MaxWorkers` parallel calls
  - after the first call, the total count is determined, and further batches are executed in parallel, if required
  - if a `getter` returns an error, the operation is aborted (no more batches are requested)
  - if a `processor` returns an error or forces to abort then the operation is aborted (no more batches are requested)
 
- Implemented specific usages of the concurrent batch processing for filter GetDimensionOptions and dataset GetOptions:
  - `filter.GetDimensionOptionsInBatches` / `dataset.GetOptionsInBatches` aggregate batches and return the full list of items
  - `filter.GetDimensionOptionsBatchProcess` / `dataset.GetOptionsBatchProcess` execute the provided batch processor for each batch (more efficient)

- Added unit tests for:
  - generic batch processing
  - filter client batch processing
  - dataset client batch processing

### How to review

- Make sure changes make sense
- Make sure unit tests pass

### Who can review

Anyone

### CODE REVIEW

**Review by CarlHembrough — CHANGES_REQUESTED** [review]
Looks good overall - I like the generic batch processing. I have added some points to check 👍

File: common/batch.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		if err != nil {
+			chErr <- err
+			abort()
+		}
+		if forceAbort {
+			abort()
+		}
+
+		// unlock
+		lockResult.Unlock()
```
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply common/batch.go:83]: good point

File: dataset/dataset.go [hunk 2]
```diff
@@ -780,6 +784,64 @@ func (c *Client) GetOptions(ctx context.Context, userAuthToken, serviceAuthToken
...
 
+// GetOptionsInBatches retrieves a list of the dimension options in concurrent batches and accumulates the results
+func (c *Client) GetOptionsInBatches(ctx context.Context, userAuthToken, serviceAuthToken, collectionID, id, edition, version, dimension string, batchSize, maxWorkers int) (opts Options, err error) {
+
+	// function to aggregate items. The first received batch will initialise the structure with items with a fixed size
+	// and items are aggregated in order even if batches are received in the wrong order (e.g. due to concurrent calls)
+	var processBatch OptionsBatchProcessor = func(batch Options) (abort bool, err error) {
+		if len(opts.Items) == 0 {
+			opts.TotalCount = batch.TotalCount
+			opts.Items = make([]Option, batch.TotalCount)
```
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply dataset/dataset.go:795]: `GetOptionsInBatches` gets all the options by aggregating the paginated responses from dataset api. In this line, we initialise the full struct that will contain all the items, adding a comment in the code to make it clearer:
```
// Function to aggregate items.
// For the first received batch, as we have the total count information, will initialise the final structure of items with a fixed size equal to TotalCount.
// This serves two purposes:
//   - We can guarantee, even with concurrent calls, that values are returned in the same order that the API defines, by offsetting the index.
//   - We do a single memory allocation for the final array, making the code more memory efficient.
```

[the remark this answers is not in the document]
Comment (DavidSubiros) [reply dataset/dataset.go:795]: Please, let me know if the comment is not clear

File: dataset/dataset.go [hunk 3]
```diff
@@ -780,6 +784,64 @@ func (c *Client) GetOptions(ctx context.Context, userAuthToken, serviceAuthToken
...
+			opts.Count = batch.TotalCount
```
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply dataset/dataset.go:796]: Same as above, this corresponds to the count for the struct that will contain all the items.

File: filter/filter.go [hunk 4]
```diff
@@ -279,6 +283,51 @@ func (c *Client) GetDimensionOptionsBytes(ctx context.Context, userAuthToken, se
...
+// GetDimensionOptionsInBatches retrieves a list of the dimension options in concurrent batches and accumulates the results
+func (c *Client) GetDimensionOptionsInBatches(ctx context.Context, userAuthToken, serviceAuthToken, collectionID, filterID, name string, batchSize, maxWorkers int) (opts DimensionOptions, err error) {
+
+	// function to aggregate items. The first received batch will initialise the structure with items with a fixed size
+	// and items are aggregated in order even if batches are received in the wrong order (e.g. due to concurrent calls)
+	var processBatch DimensionOptionsBatchProcessor = func(batch DimensionOptions) (abort bool, err error) {
+		if len(opts.Items) == 0 {
+			opts.TotalCount = batch.TotalCount
+			opts.Items = make([]DimensionOption, batch.TotalCount)
+			opts.Count = batch.TotalCount
```
[the remark this answers is not in the document]
Comment (DavidSubiros) [reply filter/filter.go:295]: Added same comment

File: dataset/dataset.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (CarlHembrough) [reply dataset/dataset.go]: I see - apologies, this was me misreading the code 👍

**Review by CarlHembrough — APPROVED** [review]
👍

**Pull request merged**